# Project 2 Part I (Core)

*Project 2: Initial Analysis and Problem Selection*

---

## Objective

Perform an initial Exploratory Data Analysis (EDA) on at least four datasets, identify and diagnose potential issues, and select a specific problem to address (regression, classification, clustering, or forecasting). Submit a repository containing the selected dataset, the initial EDA, and a description of the chosen problem.

---

# Part I: Dataset Search and Analysis

## Problem Statement

Analyze historical earthquake activity in Chile to identify geographic patterns of seismic events and support risk assessment for infrastructure planning and investment decisions.

# Dataset Description

This dataset contains records of seismic events in Chile, including the date and time of occurrence, geographic coordinates, depth, and magnitude.

## Dataset Source

In this project, we analyze the **Earthquakes in Chile** dataset, obtained from Kaggle. The dataset contains information about seismic events in Chile, including the date and time of occurrence, geographic coordinates, depth, and magnitude of each earthquake. The data was originally compiled from records published by the Chilean National Seismological Center and made available through Kaggle.

The dataset can be accessed and downloaded from:
https://www.kaggle.com/datasets/nicolasgonzalezmunoz/earthquakes-on-chile

## Dataset Information

- Number of observations: **4018**
- Number of features: **5**
- Target variable: No target variable selected, see Potential Target Variable section for more details.

## Data Dictionary

| Column      | Current Data Type | Expected Data Type | Description                                                                                                                | Example               |
| ----------- | ----------------- | ------------------ | -------------------------------------------------------------------------------------------------------------------------- | --------------------- |
| `Date(UTC)` | String (`str`)    | Datetime           | Date and time when the earthquake was recorded, expressed in Coordinated Universal Time (UTC) with second-level precision. | `2025-05-26 03:50:27` |
| `Latitude`  | Float (`float64`) | Float              | Geographic latitude of the earthquake epicenter in decimal degrees.                                                        | `-19.63`              |
| `Longitude` | Float (`float64`) | Float              | Geographic longitude of the earthquake epicenter in decimal degrees.                                                       | `-69.49`              |
| `Depth`     | Integer (`int64`) | Integer            | Depth of the earthquake hypocenter, measured in kilometers below the Earth's surface.                                      | `97`                  |
| `Magnitude` | Float (`float64`) | Float              | Magnitude of the earthquake, representing the energy released during the seismic event.                                    | `5.6`                 |

**Notes**
- Each record represents a single seismic event.
- Latitude and longitude can be combined for geospatial analysis and mapping.
- Depth and magnitude are the primary seismic characteristics of each earthquake.
- The dataset does not contain a predefined target variable, making it suitable for exploratory analysis, clustering, geospatial pattern discovery, or forecasting depending on the selected problem.

**Data Quality Observation**

Although **Date(UTC)** represents temporal information, it is currently stored as a string (`str`) rather than a datetime object. Converting this variable to a datetime format will be necessary for time-based analysis, feature engineering, and potential forecasting tasks.

---

# Data Loading and Inspection

## Import Libraries

In [119]:
# Standard library imports
import math
from pathlib import Path

# Third-party imports
import pandas as pd
import numpy as np
import plotly.graph_objects as go
import plotly.express as px
from plotly.subplots import make_subplots
import ipywidgets as widgets

### Dataset Availability and Path Setup

In [81]:
data_dir = Path("../datasets")

# List available files (for verification)
available_files = list(data_dir.iterdir())
print("Available files:", available_files)
file_name = "2_earthquakes_chile.csv"
dataset_path = data_dir / file_name

Available files: [WindowsPath('../datasets/1_aircrafts_chile_30042026.csv'), WindowsPath('../datasets/2_earthquakes_chile.csv'), WindowsPath('../datasets/3_healthy_diet.csv'), WindowsPath('../datasets/4_temp_chile_2014.csv'), WindowsPath('../datasets/README.md')]


In [82]:
# Validate dataset existence
if not dataset_path.exists():
    raise FileNotFoundError(f"{file_name} not found in {data_dir}")

print("Dataset found at:", dataset_path)

Dataset found at: ..\datasets\2_earthquakes_chile.csv


## Load Dataset

In [83]:
df = pd.read_csv(dataset_path)

## Dataset Overview


### First and Last Records

In [84]:
df.head()

,Date(UTC),Latitude,Longitude,Depth,Magnitude
0,2025-05-26 03:50:27,-19.63,-69.49,97,5.6
1,2025-05-13 00:47:58,-51.25,-72.28,28,5.1
2,2025-05-05 09:46:48,-29.49,-71.84,48,5.0
3,2025-05-05 02:17:48,-31.89,-70.88,88,5.1
4,2025-05-02 15:23:49,-27.52,-72.48,30,5.8


In [85]:
df.tail()

,Date(UTC),Latitude,Longitude,Depth,Magnitude
4013,2012-03-24 07:28:33,-33.05,-71.06,69,5.2
4014,2012-03-19 21:52:39,-25.00,-69.75,100,5.0
4015,2012-03-10 02:26:57,-19.74,-69.25,101,5.3
4016,2012-03-04 16:27:20,-21.60,-70.06,47,5.3
4017,2012-03-03 11:01:47,-30.19,-71.45,35,5.6


The dataset contains 4,018 earthquake records and 5 variables. Based on the first and last records, the data appears to be sorted in descending chronological order, covering seismic events between March 2012 and May 2025.

### Verify Structure

In [86]:
print(f"{file_name}")
print(f"Shape: {df.shape}")
print(f"→ {df.shape[0]} rows, {df.shape[1]} columns\n")

2_earthquakes_chile.csv
Shape: (4018, 5)
→ 4018 rows, 5 columns



---

## Initial Data Inspection

### Inspect Data Types

In [87]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 4018 entries, 0 to 4017
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Date(UTC)  4018 non-null   str    
 1   Latitude   4018 non-null   float64
 2   Longitude  4018 non-null   float64
 3   Depth      4018 non-null   int64  
 4   Magnitude  4018 non-null   float64
dtypes: float64(3), int64(1), str(1)
memory usage: 157.1 KB


### Data Type Assessment

The dataset contains one temporal variable (`Date(UTC)`) currently stored as a string and four numerical variables.

The `Date(UTC)` column should be converted to a datetime format during preprocessing to enable temporal analysis and feature engineering.

No missing values are visible during the initial inspection, suggesting a high level of data completeness.

In [88]:
df.columns.tolist()

['Date(UTC)', 'Latitude', 'Longitude', 'Depth', 'Magnitude']

### Descriptive Statistics

In [89]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Latitude,4018.0,-27.837778,5.961963,-62.35,-31.82,-29.595,-21.5125,-11.07
Longitude,4018.0,-70.851127,1.293387,-75.73,-71.59,-71.180,-70.0300,-58.39
Depth,4018.0,61.524639,38.276610,2.00,36.00,50.000,86.0000,624.00
Magnitude,4018.0,4.428945,0.688666,2.30,4.00,4.400,4.8000,8.40


In [90]:
df.describe(include="str").T

,count,unique,top,freq
Date(UTC),4018,4015,2019-11-01 23:43:27,2


**Analysis**

Latitude values range from `-62.35` to `-11.07`, while longitude values range from `-75.73` to `-58.39`, covering a broad geographic area consistent with Chile and surrounding seismic zones.

Earthquake depths range from `2 km` to `624 km`, with a median depth of `50 km`. The large variation in depth suggests the presence of both shallow and deep seismic events.

Earthquake magnitudes range from `2.3` to `8.4`, with an average magnitude of approximately `4.4`. Moderate-magnitude earthquakes appear to be more common than extreme seismic events.

The `Date(UTC)` column contains `4,015 unique timestamps` for `4,018 observations`, indicating that some timestamps are repeated. Further investigation is required to determine whether these represent duplicate records or separate events recorded at the same second.

---

### Missing Values

In [91]:
df.isnull().sum()

Date(UTC)    0
Latitude     0
Longitude    0
Depth        0
Magnitude    0
dtype: int64

### Duplicate Records Check

In [92]:
n_duplicated = df.duplicated().sum()

duplicate_pct = (
    n_duplicated / len(df) * 100
)

print(f"Exact duplicated rows before cleaning: {n_duplicated}")
print(f"Percentage of duplicated records: {duplicate_pct:.2f}%")

Exact duplicated rows before cleaning: 3
Percentage of duplicated records: 0.07%


The dataset contains `3 exact duplicate records`, representing approximately `0.07%` of all observations. This proportion is very small, suggesting generally good data quality. The duplicated records should be reviewed and removed during the data cleaning phase if confirmed as unintended duplicates.

---

## Distribution Analysis

### Numeric

#### Histograms

In [93]:
# Select numeric columns
numeric_columns = df.select_dtypes(
    include=np.number
).columns

# Dashboard layout
n_cols = 4
n_rows = math.ceil(len(numeric_columns) / n_cols)

# Create subplot figure
fig = make_subplots(
    rows=n_rows,
    cols=n_cols,
    subplot_titles=numeric_columns
)

# Add histograms
for i, col in enumerate(numeric_columns):
    row = (i // n_cols) + 1
    col_pos = (i % n_cols) + 1

    fig.add_trace(
        go.Histogram(
            x=df[col],
            name=col,
            nbinsx=30
        ),
        row=row,
        col=col_pos
    )

# Update layout
fig.update_layout(
    title="Interactive Histogram Dashboard — Numeric Features",
    height=300 * n_rows,
    width=1400,
    showlegend=False,
    template="plotly_white",
    bargap=0.05
)

# Descriptive axis labels
fig.update_xaxes(title_text="Feature Values")
fig.update_yaxes(title_text="Number of Observations")

# Show dashboard
fig.show()

**Analysis**

`Magnitude` shows an approximately normal distribution centered around 4.4, indicating that most recorded earthquakes are of moderate intensity, while high-magnitude events above 8.0 are rare.

The remaining variables show skewed distributions. The patterns observed in `Latitude` and `Longitude` reflect the elongated geographic shape of Chile, while `Depth` indicates that  very shallow earthquakes are more common than intermediate-depth or very deep events.

### Date Type Correction

In [94]:
# Create a working copy
df_datetime = df.copy()

# Convert date column
df_datetime["Date(UTC)"] = pd.to_datetime(
    df_datetime["Date(UTC)"]
)

df_datetime.info()

<class 'pandas.DataFrame'>
RangeIndex: 4018 entries, 0 to 4017
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Date(UTC)  4018 non-null   datetime64[us]
 1   Latitude   4018 non-null   float64       
 2   Longitude  4018 non-null   float64       
 3   Depth      4018 non-null   int64         
 4   Magnitude  4018 non-null   float64       
dtypes: datetime64[us](1), float64(3), int64(1)
memory usage: 157.1 KB


The `Date(UTC)` column was successfully converted from string format to a datetime data type, enabling temporal analysis and feature extraction.

### Date Range

In [95]:
print(df_datetime["Date(UTC)"].min())
print(df_datetime["Date(UTC)"].max())

2012-03-03 11:01:47
2025-05-26 03:50:27


The dataset covers seismic events recorded between March 2012 and May 2025, providing more than 13 years of historical earthquake data.

### Duplicate Timestamp Check

In [96]:
duplicated_timestamps = (
    df_datetime["Date(UTC)"]
    .duplicated()
    .sum()
)

print(
    f"Duplicated timestamps: {duplicated_timestamps}"
)

Duplicated timestamps: 3


### Duplicate Timestamp Inspection

In [97]:
duplicated_events = (
    df_datetime[
        df_datetime["Date(UTC)"]
        .duplicated(keep=False)
    ]
    .sort_values("Date(UTC)")
)

duplicated_events

,Date(UTC),Latitude,Longitude,Depth,Magnitude
2122,2019-03-13 21:46:10,-30.71,-71.41,42,4.1
2123,2019-03-13 21:46:10,-30.71,-71.41,42,4.1
2055,2019-04-25 09:33:32,-29.66,-71.30,68,3.8
2056,2019-04-25 09:33:32,-29.66,-71.30,68,3.8
1822,2019-11-01 23:43:27,-29.25,-71.59,40,4.2
1823,2019-11-01 23:43:27,-29.25,-71.59,40,4.2


The duplicated timestamps correspond to records with identical values across all variables, confirming the presence of exact duplicate observations rather than distinct earthquakes occurring at the same second.

These duplicated records represent approximately:

In [98]:
duplicate_pct = (
    df_datetime.duplicated().sum()
    / len(df_datetime)
    * 100
)

print(
    f"{duplicate_pct:.2f}%"
)

0.07%


### Temporal Analysis

### Earthquakes by Year

In [99]:
df_datetime["Year"] = df_datetime["Date(UTC)"].dt.year

df_datetime["Year"].value_counts().sort_index()

Year
2012     19
2013     17
2014    312
2015    470
2016    325
2017    354
2018    297
2019    479
2020    443
2021    412
2022    371
2023    305
2024    184
2025     30
Name: count, dtype: int64

In [104]:
# Count earthquakes per year
yearly_counts = (
    df_datetime["Year"]
    .value_counts()
    .sort_index()
    .reset_index()
)

yearly_counts.columns = ["Year", "Earthquakes"]

yearly_counts.head()

,Year,Earthquakes
0,2012,19
1,2013,17
2,2014,312
3,2015,470
4,2016,325


In [107]:
fig = px.bar(
    yearly_counts,
    x="Year",
    y="Earthquakes",
    text="Earthquakes",
    title="Number of Recorded Earthquakes by Year"
)

fig.update_traces(
    textposition="outside"
)

fig.update_layout(
    template="plotly_white",
    xaxis_title="Year",
    yaxis_title="Number of Earthquakes",
    height=500,
    width=1000
)

fig.update_xaxes(
    tickvals=yearly_counts["Year"]
)

fig.show()

**Analysis**

The most active seismic year in the dataset is `2019`, with `479 recorded earthquakes`, followed closely by `2015` with `470 events`.

Between 2014 and 2023, the dataset shows relatively stable seismic activity, although year-to-year fluctuations are present.

The lower counts observed in 2012 and 2025 are expected because these years are only partially covered by the dataset.

In [102]:
yearly_counts.sort_values(
    by="Earthquakes",
    ascending=False
).head()

,Year,Earthquakes
7,2019,479
3,2015,470
8,2020,443
9,2021,412
10,2022,371


### Earthquakes by Month

In [108]:
# Create a copy for temporal analysis
df_monthly = df_datetime.copy()

# Extract month
df_monthly["Month"] = df_monthly["Date(UTC)"].dt.month

In [109]:
monthly_counts = (
    df_monthly["Month"]
    .value_counts()
    .sort_index()
    .reset_index()
)

monthly_counts.columns = ["Month", "Earthquakes"]

monthly_counts

,Month,Earthquakes
0,1,307
1,2,298
2,3,324
3,4,456
4,5,306
5,6,276
6,7,289
7,8,302
8,9,465
9,10,345


In [110]:
fig = px.bar(
    monthly_counts,
    x="Month",
    y="Earthquakes",
    text="Earthquakes",
    title="Number of Recorded Earthquakes by Month"
)

fig.update_traces(
    textposition="outside"
)

fig.update_layout(
    template="plotly_white",
    xaxis_title="Month",
    yaxis_title="Number of Earthquakes",
    height=500
)

fig.update_xaxes(
    tickmode="array",
    tickvals=list(range(1, 13)),
    ticktext=[
        "Jan", "Feb", "Mar", "Apr",
        "May", "Jun", "Jul", "Aug",
        "Sep", "Oct", "Nov", "Dec"
    ]
)

fig.show()

**Analysis**

The number of recorded earthquakes is relatively stable throughout the year. September (465 events) and April (456 events) show the highest counts, although no strong seasonal pattern is evident.

The highest earthquake activity is observed in September (465 events) and April (456 events), while June records the lowest number of events (276).

Although some monthly variation exists, the overall distribution appears fairly consistent, suggesting that earthquake occurrence does not follow a strong seasonal pattern.

The difference between the highest month (465) and the lowest month (276) is noticeable, but not extreme considering the dataset contains **4,018 earthquakes over more than 13 years**. That supports the conclusion that seismic activity is fairly constant over time.

Further analysis could investigate whether earthquake magnitude or depth varies across months and whether specific regions contribute to the observed peaks.

### Countplots

#### Outlier Detection

In [111]:
# Select numeric columns
numeric_columns = df.select_dtypes(
    include=np.number
).columns

if len(numeric_columns) > 0:

    # create boxplots
    # Dashboard layout
    n_cols = 4
    n_rows = math.ceil(len(numeric_columns) / n_cols)

    # Create subplot figure
    fig = make_subplots(
        rows=n_rows,
        cols=n_cols,
        subplot_titles=numeric_columns
    )

    # Add boxplots
    for i, col in enumerate(numeric_columns):
        row = (i // n_cols) + 1
        col_pos = (i % n_cols) + 1

        fig.add_trace(
            go.Box(
                y=df[col],
                name=col,
                boxmean=True
            ),
            row=row,
            col=col_pos
        )

    # Update layout
    fig.update_layout(
        title="Interactive Boxplot Dashboard — Numeric Features",
        height=300 * n_rows,
        width=1400,
        showlegend=False,
        template="plotly_white"
    )

    # Add axis labels
    fig.update_yaxes(title_text="Observed Values")

    # Show dashboard
    fig.show()

else:
    print("No numeric variables available for outlier analysis.")

**Analysis**

All numeric variables contain observations identified as outliers according to the IQR rule.

However, the percentage of outliers is relatively low, ranging from 0.20% for `Latitude` to 2.69% for `Magnitude`.

Since these values correspond to real earthquake measurements, they should not automatically be considered errors. Extreme magnitudes, depths, or geographic locations may represent significant seismic events and could provide valuable information for the analysis.

If required for modeling, transformations such as logarithmic scaling or robust scaling could be evaluated to reduce the influence of extreme values while preserving the original observations.

IQR calculation.

In [ ]:
feature_columns = df.select_dtypes(
    include=np.number
).columns

outlier_summary = []

for col in feature_columns:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    n_outliers = ((df[col] < lower_bound) | (df[col] > upper_bound)).sum()

    outlier_summary.append({
        "Feature": col,
        "Outliers": n_outliers,
        "Percentage": round(n_outliers / len(df) * 100, 2)
    })

outlier_df = pd.DataFrame(outlier_summary)
outlier_df

,Feature,Outliers,Percentage
0,Latitude,8,0.20
1,Longitude,102,2.54
2,Depth,55,1.37
3,Magnitude,108,2.69


---

# Exploratory Data Analysis (EDA)

### Correlation Analysis

In [ ]:
numeric_columns = df.select_dtypes(
    include=np.number
).columns

corr_matrix = df[numeric_columns].corr()

corr_matrix

,Latitude,Longitude,Depth,Magnitude
Latitude,1.000000,0.727304,0.383979,0.089836
Longitude,0.727304,1.000000,0.668680,0.104494
Depth,0.383979,0.668680,1.000000,0.212480
Magnitude,0.089836,0.104494,0.212480,1.000000


### Heatmap visualization

In [ ]:
# Correlation matrix
corr_matrix = df[numeric_columns].corr()

# Create mask for upper triangle
mask = np.triu(
    np.ones_like(corr_matrix, dtype=bool)
)

# Apply mask
corr_matrix_masked = corr_matrix.mask(mask)

fig = px.imshow(
    corr_matrix_masked,
    text_auto=".2f",
    color_continuous_scale="BuGn",
    aspect="auto",
    title="Correlation Matrix Heatmap"
)

fig.update_layout(
    width=900,
    height=800
)

fig.show()

**Analysis**

`Latitude` and `Longitude` show a relatively strong positive correlation, reflecting the geographic distribution of earthquakes along Chile. `Depth` also exhibits a moderate positive correlation with `Longitude`, suggesting that earthquake depth may vary across different seismic regions.

`Latitude` has a moderate correlation with `Depth` (0.38), indicating that some regions may experience deeper earthquakes than others.

`Magnitude` shows very weak correlations with all other variables, suggesting that earthquake intensity is not strongly associated with location or depth in a simple linear relationship.

These results indicate that geographic variables may be useful for identifying spatial patterns of seismic activity, while earthquake magnitude may require more advanced analysis to uncover meaningful relationships.

## Geospatial Visualization

In [136]:
# Create copy for geospatial analysis
df_geo = df_datetime.copy()

# Slider
min_magnitude = widgets.FloatSlider(
    value=5.0,
    min=float(df_geo["Magnitude"].min()),
    max=float(df_geo["Magnitude"].max()),
    step=0.1,
    description="Min Mag:",
    continuous_update=False
)

def plot_earthquakes(min_mag):

    filtered_df = df_geo[
        df_geo["Magnitude"] >= min_mag
    ]

    filtered_df = filtered_df.copy()

    filtered_df["Marker_Size"] = (
        filtered_df["Magnitude"] - filtered_df["Magnitude"].min()
    ) * 10 + 8

    print(
        f"Displayed earthquakes: {len(filtered_df)}"
    )

    fig = px.scatter_map(
        filtered_df,
        lat="Latitude",
        lon="Longitude",
        color="Depth",
        size="Marker_Size",
        size_max=30,
        hover_data=[
            "Magnitude",
            "Depth",
            "Date(UTC)"
        ],
        color_continuous_scale="Viridis_r",
        zoom=3,
        center={
            "lat": -35,
            "lon": -71
        },
        title=f"Earthquakes in Chile (Magnitude ≥ {min_mag:.1f})"
    )

    fig.update_traces(
        marker=dict(
            sizemin=4
        )
    )

    fig.update_layout(
        template="plotly_white",
        height=700,
        coloraxis_colorbar_title="Depth (km)"
    )

    fig.show()


# Output area
output = widgets.interactive_output(
    plot_earthquakes,
    {"min_mag": min_magnitude}
)

# Show map first, controls below
display(
    widgets.VBox([
        output,
        min_magnitude
    ])
)

**Analysis**

Earthquakes are distributed along the entire Chilean territory, following the Pacific coast and the Andes region. Larger earthquakes appear concentrated in specific seismic zones, while depth varies considerably across locations. Deeper earthquakes are generally observed inland, whereas shallower events are more common near the coast and offshore areas, reflecting the subduction process between the Nazca and South American tectonic plates.

---

## Potential Target Variable

Candidate target:
- No target variable selected

Potential ML task:
- Clustering

Rationale:

The dataset contains only descriptive information about earthquake events and does not include a variable to predict. Clustering can be applied using latitude, longitude, depth, and magnitude to discover natural groupings of earthquakes. These groups may represent different seismic regions or tectonic behaviors within Chile.

---

# EDA Findings Summary

## Findings

- The dataset contains 4,018 earthquake records and 5 variables describing the location, depth, magnitude, and occurrence date of seismic events in Chile.

- No missing values were found in any column, indicating a complete dataset.

- Three exact duplicate records were identified, representing only 0.07% of the dataset. These records should be removed during data cleaning.

- The dataset covers earthquakes recorded between March 2012 and May 2025, providing more than 13 years of historical seismic information.

- Earthquake magnitudes are approximately normally distributed around 4.4, with only a small number of high-magnitude events above 8.0.

- Earthquake activity is relatively stable throughout the year, although September and April show the highest number of recorded events.

- The year 2019 contains the largest number of recorded earthquakes, followed by 2015.

- Outliers were detected in `Latitude`, `Longitude`, `Depth`, and `Magnitude`. However, these observations represent real seismic events and should be preserved rather than removed.

- Moderate positive correlations exist between geographic coordinates and earthquake depth. In particular, `Longitude` and `Depth` show a correlation of approximately 0.67, suggesting that earthquake depth may vary by geographic region.

- Geographic visualization shows that earthquakes are distributed along the Chilean territory, following the country's seismic and tectonic activity zones.

- The dataset does not contain a natural target variable for supervised learning. Therefore, clustering is an appropriate machine learning approach to identify groups of earthquakes with similar geographic and seismic characteristics.

---

# Appendix

### References

- Earthquakes in Chile Dataset, Kaggle: [Earthquakes Dataset](https://www.kaggle.com/datasets/nicolasgonzalezmunoz/earthquakes-on-chile)
- Machine Learning notebooks from the bootcamp repository (`sonda2026` branch)
- Machine Learning notebooks from the shared Assistanship Google Drive folder
- [Pandas Documentation](https://pandas.pydata.org/docs/user_guide/index.html)
- [Plotly Documentation](https://docs.plotly.com/)

### Acknowledgements
I would like to thank my instructors for their guidance, continuous support, and encouragement throughout this project.

I also acknowledge the use of AI-assisted tools for support with debugging, code review, documentation, and learning concepts during the development of the analysis. All final decisions, interpretations, and conclusions presented in this work remain my own.